_posts/2019-08-26-how-to-use-gruntwork-infrastructure-as-code-library.adoc

```diff
@@ -1399,6 +1409,272 @@ module "vpc" {
+
+The first way is to set two environment variables in the TFC workspace: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. This is perhaps the simplest approach, because all you need to do to set it up is to generate the access keys for an IAM user, assign that user appropriate permissions, and https://learn.hashicorp.com/terraform/tfc/tfc_workspaces#editing-variables[set the environment variables]. The downside is that the credentials are static. You'll need to create each workspace in advance, rather than relying on the implicit workspace creation, and for each workspace, you'll have to set the environment variables manually. If you rotate the keys, you'll have to update them in each workspace.
+
+A second way is to generate a `provider` block in the root `terragrunt.hcl`, and include the static API credentials in the block. For example:
+
+
+[source,hcl]
+----
+locals {
+  region_vars  = read_terragrunt_config(find_in_parent_folders("region.hcl"))
+  access _key  = "<value for access key>"
+  secret_key  = "<value for secret key>"
+}
+
+generate "aws_provider" {
+  path      = "provider.tf"
+  if_exists = "overwrite"
+  contents = <<-EOF
+provider "aws" {
+  region     = "${local.region_vars.locals.aws_region}"
+  access_key = "${local._access_key}"
+  secret_key = "${local.secret_key}"
+}
+EOF
+}
+----
+
+Using this approach, the workspace does not need to be created in advance, and the environment variables do not need to be set. It's more scalable, but now the API keys are kept in the configuration in plain text. Because this configuration is stored in version control, this approach is not recommended.
+
+A third way is to generate the provider block, but to retrieve dynamic credentials using the https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html:[AWS AssumeRole feature] and the Terragrunt https://terragrunt.gruntwork.io/docs/reference/built-in-functions/#run_cmd[`run_cmd` function]. Using this approach, the `run_cmd` invokes an external script to assume an IAM role, fetch credentials, and return them to Terragrunt, where they are subsequently inserted in to the provider configuration.
+
+
+[source,hcl]
+----
+locals {
+  region_vars  = read_terragrunt_config(find_in_parent_folders("region.hcl"))
+  profile_name = "<your AWS profile>"
+  role_arn     = "<your IAM role>"
+  creds        = split(",", chomp(run_cmd("bin/get_aws_creds.py",
+                       "--profile-name",
+                       "${local.profile_name}",
+                       "--role-arn",
+                       "${local.role_arn}")))
+  access_key     = local.creds[0]
+  secret_key     = local.creds[1]
+  session_token  = local.creds[2]
+}
+
+generate "aws_provider" {
+  path      = "provider.tf"
```

Comment:
There are a few major drawbacks with all these approaches that use `generate` to create a `provider.tf` file:

1. You end up with that generated file on your local disk, with your creds in plaintext.
1. Worse yet, that file is _also_ uploaded to TFC/TFE, and stored there in plaintext.

When you add env vars in the TFC/TFE UI, you can at least mark them as "sensitive," so they are stored in an encrypted format. 

The biggest issue of all, however, is that, AFAIK, 90% of the reason to use TFC/TFE as a remote backend is so that developers do NOT need to have AWS creds of any kind locally. For big companies, distributing such creds is a messy affair, and many of them use TFC/TFE specifically to avoid that.

Therefore, my vote is that we recommend the following:

1. Create your workspaces first. An admin can do this manually... Or, perhaps you set up your Terragrunt folder structure, and run `validate-all`. I _think_ that will run `init` in each module, which will create all the workspaces.
1. Have an admin provide machine user AWS credentials for each workspace. E.g., With TFE, you could assign an IAM role to the server. With TFC, the admin may have to go into each workspace and set the access keys as env vars in the UI, marking them as "sensitive." One other option for admins could be to use the [tfe_workspace_ids data source](https://www.terraform.io/docs/providers/tfe/d/workspace_ids.html) to periodically retrieve all workspaces (set `names` to `*`) and then loop over them using `for_each` in a [tfe_variable resource](https://www.terraform.io/docs/providers/tfe/r/variable.html) to add the AWS access keys as env vars.
1. Once all workspaces are configured, now the devs can run `terragrunt xxx` and it will execute in that workspace with the credentials it needs.

Reply:
> There are a few major drawbacks with all these approaches that use generate to create a provider.tf file:

At least in the case of the `run_cmd` approach the credentials are temporary. The script could generate credentials that are only valid for 15 minutes. Seems like a reasonable balance to me, but admittedly it isn't a super clean solution. I think we should at least note it as an option.

> The biggest issue of all, however, is that, AFAIK, 90% of the reason to use TFC/TFE as a remote backend is so that developers do NOT need to have AWS creds of any kind locally. For big companies, distributing such creds is a messy affair, and many of them use TFC/TFE specifically to avoid that.

These seems anecdotal to me. It may be one reason, but it may equally be that they want to track the runs centrally, or use TFC/TFE for storing state, or want sentinel... or many other reasons. I honestly can't imagine trying to be a developer trying to do stuff in the cloud without any kind of credentials said cloud. But I suppose from a separation of duties standpoint, a prod environment might be read-only for the developer, in which case the approach described above would be a concern. But then why not just have whomever the admin is run terraform?

I don't see manually creating workspaces first as a practical solution. The workspaces are likely to change often. Seems pretty error prone. But probably worth writing up as another alternative.

Yet another option would be a wrapper for terragrunt! It could request temporary AWS credentials, put them in a place where terraform can reach them (either in the workspace using the `tfe_variable` resource, as you mentioned, or in a `*.auto.tfvars` file, in a terragrunt.hcl... numerous ways to do it), then invoke terragrunt, and finally clean up after itself when complete.

However, this extra layer of abstraction seems a little hacky too.